text/0002-serializable-types.md

```diff
@@ -0,0 +1,413 @@
+
+```cs
+[ClientRpc]
+void BarClientRpc(Color somecolor) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        BarClientRpc(Color.red); // Server -> Client
+    }
+}
+```
+
+### Enum Types
+
+A user-defined enum type will be serialized by built-in serialization code (with underlying integer type).
+
+```cs
+enum SmallEnum : byte
+{
+    A,
+    B,
+    C
+}
+
+enum NormalEnum // default -> int
+{
+    X,
+    Y,
+    Z
+}
+
+[ServerRpc]
+void ConfigServerRpc(SmallEnum smallEnum, NormalEnum normalEnum) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        ConfigServerRpc(SmallEnum.A, NormalEnum.X); // Client -> Server
+    }
+}
+```
+
+### Static Arrays
+
+Static arrays like `int[]` will be serialized by built-in serialization code if their underlying type is either one of serialization supported types (e.g. `Vector3`) or if they implement `INetworkSerializable` interface.
+
+```cs
+[ServerRpc]
+void HelloServerRpc(int[] scores, Color[] colors) { /* ... */ }
+
+[ClientRpc]
+void WorldClientRpc(MyComplexType[] values) { /* ... */ }
+```
+
+### NetworkObject & NetworkBehaviour
+
+`NetworkObject` and `NetworkBehaviour` instances will be serialized by built-in serialization code if instances are not `null` and `NetworkObject.IsSpawned == true`. Passing a `null` instance or a `NetworkObject` with `IsSpawned == false` or a `NetworkBehaviour` with `NetworkObject.IsSpawned == false` will still lead into RPC replication but remote side will have `null` values for the parameters. IDs of spawned `NetworkObject` and `NetworkBehaviour` instances will be resolved by running `NetworkManager` instance, and those IDs will be the links between local and remote instances. Also, those IDs will be used when serializing `NetworkObject` and `NetworkBehaviour` instances as a part of an RPC call.
+
+```cs
+[ServerRpc]
+void MyServerRpc(NetworkObject nobj, NetworkBehaviour nbhv) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        MyServerRpc(/* NetworkObject = */ this.NetworkObject, /* NetworkBehaviour = */ this); // Client -> Server
+    }
+}
+```
+
+### INetworkSerializable & BitSerializer
+
+Complex user-defined types that implements `INetworkSerializable` interface will be serialized by user provided serialization code.
+
+An instance of `BitSerializer` will be passed into `INetworkSerializable::NetworkSerialize(BitSerializer)` method which can be used to easily serialize fields by reference.
+
+All types supporting serialization will also be supported by `BitSerializer` with `BitSerializer::Serialize(ref int value)` variant methods and templated `BitSerializer::Serialize<T>(ref T value) where T : INetworkSerializable` method.
+
+```cs
+class BitSerializer
+{
+    bool IsReading { get; }
+
+    void Serialize(ref int value) { /* ... */ }
```

Comment:
True / False: in the end, all reads/writes to BitSerializer will be done through one of the built-in types.  From the examples and my understanding, even if I have a custom type, my serialization operations will just be a sequence of built-in type operations.

If I have this right, it would mean you could not, say, do weird bitwise packing (like pack a bool into a byte or even a bit).  I'm not saying that's a bad thing, IIRC @pandr & his group found this to be not worth the bandwidth savings.  But I would have liked this called out earlier on to help me get that insight sooner.

If I have this wrong, I'd suggest showing an example (presumably like how you'd get to m_writer in BitWriter).

Reply:
yes, you got it right but there are some other things worth mentioning here. `BitSerializer` will not expose `BitWriter` and `BitReader` it holds internally but it will expose `Serialize(ref byte value)` and `Serialize(ref byte[] value)` API which could be used for any kind of manual serialization including bit packing, quantization and other compression techniques. I think I should inline the whole `BitSerializer` API here to clarify what's being proposed here (looks like some things stuck in my mind but not written entirely here)

Reply:
Great, yeah I think just more or less pasting in what you just described would be great just to indicate you can do such trickery